### ISSUE TITLE
debug console with API connection sometimes hangs since 2.8.2

### ISSUE THREAD

**Issue by dupondje** [Icinga/icinga2#6184]
Hi All,

I use some easy script that shows a diff of checks when doing a config change in icinga2. So that its visible what checks were added/removed when adjusting the config.

All it does is:
`ICINGA2_API_PASSWORD=xxxx icinga2 console --connect 'https://root@localhost:5665/' --eval "get_objects(Service).map(s => s.__name).sort();"`

Now since upgrading to 2.8.2, we've noticed this command seems to hang quite often.
Even if we run it manually, we see it hanging, no output comes from icinga2, and not errors in the logs.

I've enabled debug logging, and no errors are thrown there neither. The only issue I can see is the following:
When it works I see the following:
```
[2018-03-26 12:18:59 +0200] information/ApiListener: New client connection from [127.0.0.1]:47054 (no client certificate)
[2018-03-26 12:18:59 +0200] notice/ApiListener: New HTTP client
[2018-03-26 12:18:59 +0200] debug/HttpRequest: line: POST /v1/console/execute-script?command=get_objects%28Service%29.map%28s%20%3D%3E%20s.__name%29.sort%28%29%3B&sandboxed=0&session=mon001-1522059539-0 HTTP/1.1, tokens: 3
[2018-03-26 12:18:59 +0200] information/HttpServerConnection: Request: POST /v1/console/execute-script?command=get_objects%28Service%29.map%28s%20%3D%3E%20s.__name%29.sort%28%29%3B&sandboxed=0&session=mon001-1522059539-0 (from [127.0.0.1]:47054, user: root)
[2018-03-26 12:18:59 +0200] notice/WorkQueue: Spawning WorkQueue threads for 'HttpServerConnection'
[2018-03-26 12:18:59 +0200] notice/Console: Executing expression: get_objects(Service).map(s => s.__name).sort();
[2018-03-26 12:18:59 +0200] debug/HttpServerConnection: Http client disconnected
[2018-03-26 12:18:59 +0200] notice/WorkQueue: Stopped WorkQueue threads for 'HttpServerConnection'
```

When it hangs:
```
[2018-03-26 12:19:06 +0200] information/ApiListener: New client connection from [127.0.0.1]:47072 (no client certificate)
[2018-03-26 12:19:06 +0200] notice/ApiListener: New HTTP client
[2018-03-26 12:19:06 +0200] debug/HttpRequest: line: POST /v1/console/execute-script?command=get_objects%28Service%29.map%28s%20%3D%3E%20s.__name%29.sort%28%29%3B&sandboxed=0&session=mon001-1522059546-0 HTTP/1.1, tokens: 3
[2018-03-26 12:19:06 +0200] information/HttpServerConnection: Request: POST /v1/console/execute-script?command=get_objects%28Service%29.map%28s%20%3D%3E%20s.__name%29.sort%28%29%3B&sandboxed=0&session=mon001-1522059546-0 (from [127.0.0.1]:47072, user: root)
[2018-03-26 12:19:06 +0200] notice/WorkQueue: Spawning WorkQueue threads for 'HttpServerConnection'
```

```
[2018-03-26 12:19:16 +0200] information/WorkQueue: #28 (HttpServerConnection) items: 0, rate: 0.0166667/s (1/min 1/5min 1/15min);
[2018-03-26 12:19:16 +0200] information/HttpServerConnection: No messages for Http connection have been received in the last 10 seconds.
[2018-03-26 12:19:16 +0200] debug/HttpServerConnection: Http client disconnected
[2018-03-26 12:19:16 +0200] information/HttpServerConnection: Unable to disconnect Http client, I/O thread busy
[2018-03-26 12:19:16 +0200] notice/WorkQueue: Stopped WorkQueue threads for 'HttpServerConnection'
```

Lets hope this can be fixed!

**Comment by lgraziosi** [comment]
Same problem here. Since 2.8.2 every command on icinga2 console hangs.

**Comment by Crunsher** [comment]
I see, when we implemented out fix we did not have the remote console on our mind. Ie. somebody having a good reason to keep the connection open for >10 seconds without doing anything. :G

**Comment by dupondje** [comment]
@Crunsher: Thats not the problem I guess. Its not that I'm waiting 10 seconds to execute it ..

I run the following from shell:
`ICINGA2_API_PASSWORD=xxxx icinga2 console --connect 'https://root@localhost:5665/' --eval "get_objects(Service).map(s => s.__name).sort();"`

And that hangs like 1 out of 3 times.

**Comment by dnsmichi** [comment]
How many (service) objects? get_objects(Service) is pretty expensive.

**Comment by dupondje** [comment]
2929 :)

Thing is, most of the time its quite fast (within the second). And other moments it just hangs. And according to the logs I pasted before, it doesn't even seem to get executed.

**Comment by dnsmichi** [comment]
If you catch such a case, can you please extract a full gdb backtrace via pid connect? Details in https://www.icinga.com/docs/icinga2/latest/doc/21-development/#gdb-backtrace-from-running-process

**Comment by dupondje** [comment]
[gdb-icinga2.txt](https://github.com/Icinga/icinga2/files/1848377/gdb-icinga2.txt)

Attached a gdb output when the command was hanging.

**Comment by Crunsher** [comment]
There is no timeout and the client waits for the server indefinitely. I am still unsure why those requests sometimes take forever, it's only happening sporadically.

The request shouldn't take that long but there are cases where this could very well happen and we don't want to kill a command just because it takes a while. But since the server will kill the connection after 10 seconds anyways, this sounds like a good timeout for the client.

This would solve the problem with it hanging.

**Comment by dnsmichi** [comment]
check_nscp_api has the same problem.

**Comment by Crunsher** [comment]
I'm tempted to call wont-fix for the nscp issue since we want to move away from the icinga http client in that case. Is that a common issue? I didn't run into that when testing #6207

**Comment by dnsmichi** [comment]
On the long run, yes, maybe. For the short run, we cannot throw away check_nscp_api in 2.9 I believe. That being said, I want to release 2.9 quite soon after resolving remaining issues and implementing customer sponsored features.

I can reproduce the issue in 2 out of 5 queries. It might be the case that NSClient++ itself behaves wonky, but looking at the lldb stacktrace, the handler is waiting forever. I'll debug further, since the code is the exact same as the console command.

**Comment by Crunsher** [comment]
It's a problem with the underlying low level logic in the http client library. We can solve both problems with one patch, it's just someone has to write that patch > <

**Comment by dupondje** [comment]
Something broke on server side between 2.8.1 and 2.8.2. We never noticed a single hang in 2.8.1, but since upgrading to 2.8.2, like 2 out of 5 api queries hang indeed.

Correct that the client should have some timeout, but the server-side introduced some bug between 2.8.1 and 2.8.2 I think!

**Comment by dnsmichi** [comment]
I definitely is a client problem, as check_nscp_api talks to NSClient++ and not Icinga 2.

**Comment by Crunsher** [comment]
We changed quite a bit for the security issues in 2.8.2 with the http server, it's possible this just made the issue visible.

**Comment by dnsmichi** [comment]
It is somewhere in `HttpClientConnection::DataAvailableHandler()` and SocketEvents where poll just hangs indefinitely. The callback for completion for the request is never called.

**Comment by dnsmichi** [comment]
Happens inside m_Stream->Close() in DataAvailableHandler(). I've seen this before with holding a reference on its own.

```
(lldb) r
Process 4863 launched: '/usr/local/sbin/check_nscp_api' (x86_64)
Sending request to 'https://10.10.98.87:8443/query/check_nscp'
[2018-04-09 15:44:17 +0200] critical/HttpClientConnection: DataAvailableHandler: Before m_Stream->IsEof().
[2018-04-09 15:44:17 +0200] critical/HttpClientConnection: DataAvailableHandler: After m_Stream->IsEof().
[2018-04-09 15:44:17 +0200] critical/HttpClientConnection: DataAvailableHandler: Closing socket.
```

**Comment by dnsmichi** [comment]
[check_nscp_lldb_bt_steps.zip](https://github.com/Icinga/icinga2/files/1890193/check_nscp_lldb_bt_steps.zip)

**Comment by dnsmichi** [comment]
Haven't found a good solution or indication here, this all is _very_ asynchronous and a race condition. DataAvailableHandler() is only called if there's data and the stream is not EOF. Inside the function call, the stream is EOF and nothing is processed anymore (no call to ProcessMessage()) and as such the call just hangs indefinitely. I've tested this, it also happens with v2.8.1 but highly likely not so often. It may be the case that "corked" influences this thing.

I'm inclined to entirely drop the httpclientconnection thing, and just implement a light-weight variant with synchronous request from the InfluxDBWriter code. Though this doesn't help the console where longer script expressions may run into timeouts, it will only help check_nscp_api running standalone against NSClient++.

Still I'd love to find out why m_Stream->IsEof() is true aka why another IO thread is interfering here.

**Comment by ekeih** [comment]
Hi,

do I understand this issue correctly that it is a client implementation problem in `icinga2 console` and not an API bug?
So it should be safe to update to 2.8.2 (or 2.8.4) when we use the API via HTTP?

Thank you in advance!

**Comment by Crunsher** [comment]
@ekeih Yes, this is only about the `icinga2 console` command. The http API is not affected.

**Comment by Crunsher** [comment]
I'm pausing this for now. I rewrote the whole ApiClient twice but the issue is with the server too. After `HttpRequest::ParseBody()`, `HttpServerConnection::ProcessMessage()` is not called again (in some cases). I failed to understand as to why though and why it would work with curl.

**Comment by Al2Klimov** [comment]
As discussed w/ @Crunsher (offline):

I'll try Boost ASIO. If it succeeds, it will be an option for problems like #6361 in the future.

**Comment by Al2Klimov** [comment]
Please don't use the last commit in production. Thanks. (However it seems to fix the problem.)

**Comment by dnsmichi** [comment]
# Summary

There's no way we can use boost::asio or any other client library. That's an open end, and won't help here.

# Problem Analysis

We've added debug logging to each stage within

- HttpServerConnection::DataAvailableHandler()
- HttpServerConnection::ProcessMessage()
- HttpRequest::ParseBody()


## Chunked Encoding

Our HTTP clients use chunked encoding by default (this is hardcoded if HTTP/1.1 is selected).

> curl does not use this by default which is why the error is **not** visible with that test method.


## Flow

Data Available -> Process Message Loop -> Body not complete -> Parse Body (Chunked)

```
if (Headers->Get("transfer-encoding") == "chunked") {
```

calls `HttpChunkedEncoding::ReadChunkFromStream` which reads the chunked body.

> Imagine sending just an empty JSON body `{}` with the size of 2 characters.

### 1. Iteration

ParseBody() returns true, since size is not 0. This returns `{}` as data. `CompleteBody` is set to true, there's no more data expected afterwards.

### 2. Iteration

ParseBody() returns false, since size == 0, no more chunked data to read. 

## The Problem

When ParseBody() returns false in the last call, this is returned to ProcessMessage().

```
while (ProcessMessage())
```

This immediately breaks the loop. We never will jump into ProcessMessage() again.

```
if (!m_CurrentRequest.CompleteBody) {
```

gets skipped, to call the code parts below. 

```
        m_RequestQueue.Enqueue(std::bind(&HttpServerConnection::ProcessMessageAsync,
                HttpServerConnection::Ptr(this), m_CurrentRequest, response, m_AuthenticatedUser));
```

## The Solution

Once `CompleteBody` is completed, ParseBody() needs to return true to indicate the caller that one more ProcessMessage() call is needed. #6383 fixes a regression introduced in 2.8.2 in ee5954726d.
